.github/scripts/check_manifest_features.py

```diff
@@ -0,0 +1,223 @@
+ * Human-written fields are not empty
+"""
+
+import glob
+import json
+from os.path import basename, dirname, splitext
+import tla_utils
+from tree_sitter import Language, Parser
+
+# Builds the tree-sitter-tlaplus grammar and constructs the parser
+Language.build_library(
+  'build/tree-sitter-languages.so',
+  ['tree-sitter-tlaplus']
+)
+
+TLAPLUS_LANGUAGE = Language('build/tree-sitter-languages.so', 'tlaplus')
+parser = Parser()
+parser.set_language(TLAPLUS_LANGUAGE)
+
+def parse_module(path):
+    """
+    Parses a .tla file; returns the parse tree along with whether a parse
+    error was detected.
+    """
+    module_text = None
+    with open(path, 'rb') as module_file:
+        module_text = module_file.read()
+    tree = parser.parse(module_text)
+    return (tree, module_text, tree.root_node.has_error)
+
+def get_module_names_in_dir(dir):
+    """
+    Gets the module names of all .tla files in the given directory.
+    """
+    return set([splitext(basename(path))[0] for path in glob.glob(f'{dir}/*.tla')])
+
+# This query looks for pluscal and proof constructs
+# It can be extended to look for other things if desired
+feature_query = TLAPLUS_LANGUAGE.query(
+    '(pcal_algorithm_start) @pluscal'
+    + '[(terminal_proof) (non_terminal_proof)] @proof'
+)
+
+def get_tree_features(tree):
+    """
+    Returns any notable features in the parse tree, such as pluscal or proofs
+    """
+    return set([name for _, name in feature_query.captures(tree.root_node)])
+
+def get_module_features(path):
+    """
+    Gets notable features for the .tla file at the given path
+    """
+    tree, _, _ = parse_module(path)
+    return get_tree_features(tree)
+
+# Keywords mapping to features for models
+model_features = {
+    'PROPERTY': 'liveness',
+    'PROPERTIES': 'liveness',
+    'SYMMETRY': 'symmetry',
+    'ALIAS': 'alias',
+    'VIEW': 'view',
+    'CONSTRAINT': 'state constraint',
+    'CONSTRAINTS': 'state constraint',
+}
+
+def get_model_features(path):
+    """
+    Finds features present in the given .cfg model file.
+    This will be a best-effort text search until a tree-sitter grammar is
+    created for .cfg files.
+    """
+    features = []
+    model_text = None
+    with open(path, 'rt') as model_file:
+        model_text = model_file.read()
+    for line in model_text.split('\n'):
+        tokens = line.split()
+        if len(tokens) > 0 and tokens[0] in model_features:
+            features.append(model_features[tokens[0]])
+    return set(features)
+
+# Query finding all module imports in a .tla file
+import_query = TLAPLUS_LANGUAGE.query(
+    '(extends (identifier_ref) @extends)'
+    + '(instance (identifier_ref) @instance)'
+)
+
+# Query finding all defined module names in a .tla file
+# Especially useful for finding nested module names
+module_name_query = TLAPLUS_LANGUAGE.query(
+    '(module name: (identifier) @module_name)'
+)
+
+# All the standard modules available when using TLC
+tlc_modules = {
```

Comment:
The module lists below are not static.  What will happen if the lists become outdated?

Reply:
Then you change it, this is a best-effort method of getting the community dependencies. The only alternative is self-reporting.

Reply:
My question is how we can help whoever will diagnose this problem in the future.  Something like a comment that says "if you see error XYZ in the CI output, add Frob to the list of modules below".

Reply:
There won't be any sort of error, all that will happen is the `communityModules` list will be incorrect in the json. This field is for the benefit of humans who might be interested in community module usage and isn't consumed by any tools.